README.md
# browserify-middleware

<img src="http://i.imgur.com/6cyfaYS.png" align="right" />

**middleware for browserify v2 with sensible defaults for the ultimate in ease of use**

In addition to the basics, browserify-middleware has the following features out of the box:

 - source-maps are automatically enabled for debugging
 - automatically rebuilds whenever files change in development
 - minification automatically enabled for production
 - gzip automatically enabled for production
 - etags for caching automatically enabled for production

With the exception of serving up directories (which requires `req.path` from express) everything is entirely framework independent.  Simply pass in `req` `res`, and a `callback` that will only be called in the event of an error.

If you think I've missed something, be sure to open an issue or submit a pull request.

[![Build Status](https://img.shields.io/travis/ForbesLindesay/browserify-middleware/master.svg)](https://travis-ci.org/ForbesLindesay/browserify-middleware)
 [![Coverage Status](https://img.shields.io/coveralls/ForbesLindesay/browserify-middleware/master.svg?style=flat)](https://coveralls.io/r/ForbesLindesay/browserify-middleware?branch=master)
[![Dependency Status](https://img.shields.io/david/ForbesLindesay/browserify-middleware.svg)](https://david-dm.org/ForbesLindesay/browserify-middleware)
[![NPM version](https://img.shields.io/npm/v/browserify-middleware.svg)](https://www.npmjs.com/package/browserify-middleware)

<a target='_blank' rel='nofollow' href='https://app.codesponsor.io/link/gg9sZwctSLxyov1sJwW6pfyS/ForbesLindesay/browserify-middleware'>
  <img alt='Sponsor' width='888' height='68' src='https://app.codesponsor.io/embed/gg9sZwctSLxyov1sJwW6pfyS/ForbesLindesay/browserify-middleware.svg' />
</a>

## Usage

See `example` directory for a complete server

```javascript
var browserify = require('browserify-middleware');
var express = require('express');
var app = express();

//provide browserified versions of all the files in a directory
app.use('/js', browserify(__dirname + '/client/dir'));

//provide a browserified file at a path
app.get('/js/file.js', browserify(__dirname + '/client/file.js'));

//provide a bundle exposing `require` for a few npm packages.
app.get('/js/bundle.js', browserify(['hyperquest', 'concat-stream']));

//provide a bundle for a few npm packages plus run main.js
app.get('/js/bundle.js', browserify(['hyperquest', 'concat-stream', {__dirname + '/client/main.js': {run: true}}]));

app.listen(3000);
```
## Multiple Bundles Example

Multiple bundles can sometimes lead to better caching performance.  If you had multiple different JavaScript modules in `./client` that all depended on `hyperquest` and `concat-stream` and were used on different pages, you may want to split those two modules into separate files so that they are only loaded once for someone browsing arround the site:

```javascript
var shared = ['hyperquest', 'concat-stream'];
app.get('/js/bundle.js', browserify(shared));
app.use('/js', browserify('./client', {external: shared}))
```

Then on your HTML pages you can just have:

page1.html
```html
<script src="/js/bundle.js"></script>
<script src="/js/beep.js"></script>
```

page2.html
```html
<script src="/js/bundle.js"></script>
<script src="/js/boop.js"></script>
```

This way, booth `beep.js` and `boop.js` can `require` the shared modules (`hyperquest` and `concat-stream`) but they aren't actually contained within that file.

## API

### `browserify('./path/to/file.js'[, options])`

Return the middleware to serve a browserified version of the file.  The file path is relative to the calling module, not to `process.cwd()`.

### `browserify('./path/to/directory/'[, options])`

Return the middleware to serve a browserified version of all the files in a directory.  The directory path is relative to the calling module, not to `process.cwd()`.

### `browserify(['module-a', 'module-b'][, options])`

Return middleware that will expose `require` for each of the modules in the array.  This will work even if those modules are also in the `external` array.

#### `browserify([{'module-d': {expose: 'dee'}}][, options])`

Require `module-d` with custom options (to be passed on to browserify).  In this case `module-d` will be exposed as `dee`.  This can be mixed and matched with plain strings.  Note that these modules must not appear in the `external` array.

### `options` / `settings`

The `options` passed to each middleware function override the defaults specified in `settings`.

Setings has two properties `settings.production` and `settings.development` which specify the default settings for each environment.  The current environment is specified by `settings.mode` and defaults to `process.env.NODE_ENV || 'development'`

Production defaults:

```javascript
production.cache = true; // equivalent to "public, max-age=60"
production.precompile = true;
production.minify = true;
production.gzip = true;
production.debug = false;
```

To update:

```javascript
browserify.settings.production('cache', '7 days');
```

Development defaults:

```javascript
development.cache = 'dynamic';
development.precompile = false;
development.minify = false;
development.gzip = false;
development.debug = true;
```

To update:

```javascript
browserify.settings.development('gzip', true);
```

The following defaults are the same for production and development:

```javascript
external = [];
ignore = [];
ignoreMissing = false;
transform = [];
insertGlobals = false;
detectGlobals = true;
standalone = false;
grep = /\.js$/
```

To update:

```javascript
browserify.settings('external', ['hyperquest']);
//or
browserify.settings({
  ignoreMissing: true,
  insertGlobals: true,
  transform: ['rfileify']
});
```

Custom Environments:

You can also create a new custom environment:

```javascript
var test = browserify.settings.env('test');
test('minify', true);
//or
test({
  debug: true
});
```

#### cache

The cache setting determines how long content can be cached in the client's web browsers (and any caching proxies) and whether or not to cache bundles server side.  Any value other than `false` will result in them being cached server side.  The `'dynamic'` cache option is special.  It works like watchify and only re-compiles the files that have changed.  This is the fastest option for development.  It does not enable any client side caching.

If cache is `true` the client will recieve Cache Control of `"public, max-age=60"`, which caches for 60 seconds.

If cache is a `string` in the form accepted by [ms](https://npmjs.org/package/ms) it becomes: `"public, max-age=" + (ms(cache)/1000)`

If cache is a `number`, it is treated as being in milliseconds so becomes: `"public, max-age=" + (cache/1000)`

If cache is an `object` of the form `{private: true || false, maxAge: '10 minutes'}` it becomes the apropriate string.

If cache is any other `string` it will be sent directly to the client.

**N.B.** that if caching is enabled, the server never times out its cache, no matter what the timeout set for the client.

#### precompile

The precompile setting enables bundles to be precompiled/built and readily cached immediately on server startup. This option is not available when using browserify with a directory.  If `precompile` is set to `true`, the bundle will be compiled & cached at server start.

```javascript
// Precompile a browserified file at a path
app.get('/js/file.js', browserify('./client/file.js', {
  cache: true,
  precompile: true
}));

// Precompile a bundle exposing `require` for a few npm packages.
app.get('/js/bundle.js', browserify(['hyperquest', 'concat-stream'], {
  cache: true,
  precompile: true
}));
```

**N.B.**  It only makes sense to use precompiling when caching is enabled. If caching is disabled, no precompiling will happen.

#### minify

If `minify` is `true`, UglifyJS will be used to minify the resulting code.  This is `true` by default in production.  If you set it to an object, the object will be passed to uglify-js as [options](https://github.com/mishoo/UglifyJS2#the-simple-way):

 - `warnings` (default `false`) - pass `true` to display compressor warnings
 - `mangle` (default `true`) - pass `false` to skip mangling names
 - `output` (default `null`) - pass an object to specify additional [output options](http://lisperator.net/uglifyjs/codegen). The defaults are optimized for best compression.
 - `compress` (default `{}`) - pass `false` to skip compressing entirely.  Pass an object to specify custom [compressor options](http://lisperator.net/uglifyjs/compress).

#### gzip

If `gzip` is `true`, GZip will be enabled when clients support it.  This increases the memory required for caching by aproximately 50% but the speed boost can be considerable.  It is `true` by default in production.

#### debug

If `debug` is `true`, a source map will be added to the code.  This is very useful when debugging.  `debug` is `false` in production.

#### basedir

If `debug` is `true` you can provide a `string` pathname for basedir and the paths of your files in the source-map will be displayed relative to that file.  This is great for hiding the details of your local file system or tidying up the debugging of a large app.

#### plugins

An array of objects of the form `{plugin: 'name', options: {object}}`.

#### grep

The regular expression, something like [`/\.(?:js|coffee|ls)$/`](http://tinyurl.com/pawk7cu), that a filename must pass to be served using browserify from a directory.

#### hooks

There are a number of hooks that you can implement to modify the source at a few stages of processing.

e.g.

```js
app.get('/index.js', browserify('/index.js', {
  preminify: function (source) {
    return angularJsMinifier(source);
  }
}));
```

The available hooks are currently:

 - postcompile - fires after compilation, but before any minfication/gzipping
 - preminify - fires before minfication (but only if minify is enabled)
 - postminify - fires after minfication (but only if minify is enabled)

The main use case you might have for this would be adding extra minfication steps that are able to make additional assumptions about your code.  These hooks can return either a string or a Promise for a string.

#### Others

The remaining settings are all passed through to browserify, you should look at [the browserify readme](https://github.com/substack/node-browserify) if you want to know more:

- `options.external` - an array of module names that will be required from external bundles (see [browserify/multiple bundles](https://github.com/substack/node-browserify#multiple-bundles)) (default: `[]`)
- `options.ignore` - an aray of module names that are prevented from showing up in the output bundle (default: `[]`)
- `options.ignoreMissing` - set to `true` to ignore errors when a module can't be found (default: `false`).
- `options.transform` - an array of transforms to transform top level modules (default: `[]`). Each item can be:
    - `"transform-name"` - the npm name of the transform
    - `transformFunction` - the transform function
    - `["transform-name" | tranformFunction, {option1: true, ...}]` - the transform and some options
- `options.insertGlobals` - set to true to always insert `process`, `global` etc. without analysing the AST for faster builds but larger bundles (Note that `options.minify` may cause the globals to be removed again anyway) (default: false)
- `options.detectGlobals` - set to false to skip adding `process`, `global` etc.  Setting this to false may break more npm modules (default: true).
- `options.noParse` - an array of module names that should not be parsed for `require` statements of node.js style globals, can speed up loading things like jQuery that are huge but never use `require`.
- `options.standalone` - generate a standalone build (in a [umd](https://github.com/ForbesLindesay/umd) wrapper) with this name, you probably don't want this.
- `options.extensions` - an array of optional extra extensions for the module lookup machinery to use when the extension has not been specified. By default browserify considers only `.js` and `.json` files in such cases.
- `options.resolve` - lets you override the default resolution algorithm (e.g. use browserify to compile component modules)
- `options.basedir` - this shouldn't be needed as browserify-middleware already resolves to absolute paths.

You can optionally pass a single item instead of an array to any of the options that take an array.

## License

  MIT

  If you find it useful, a donation via [gittip](https://www.gittip.com/ForbesLindesay) would be appreciated.


History.md
8.1.1 / 2018-06-01
==================

  * Update ms (limits `cache` option to 100 chars if string)

8.1.0 / 2018-03-13
==================

  * Update browserify to 16.0.0
  * Swap from uglify-js to uglify-es

8.0.0 / 2017-10-18
==================

  * Update browserify to [14.1.0](https://github.com/browserify/browserify/blob/master/changelog.markdown#1410)
  * Pass other options to browserify

7.1.0 / 2016-09-19
==================

  * Update dependencies

7.0.0 / 2015-07-22
==================

  * Update browserify to [11.0.0](https://github.com/substack/node-browserify/blob/master/changelog.markdown#1100)
  * Add `bundleExternal` option (thanks to [@59naga](https://github.com/59naga))

6.0.0 / 2015-06-18
==================

  * Update browserify to [10.2.4](https://github.com/substack/node-browserify/blob/master/changelog.markdown#1024)
  * Use watchify for dynamic caching

5.0.1 / 2015-02-13
==================

  * Fix dynamic caching
  * Add code coverage to tests

5.0.0 / 2015-02-07
==================

  * Update browserify to [8.1.3](https://github.com/substack/node-browserify/blob/master/changelog.markdown#813)
  * Fix support for some boolean options
  * Add `basedir` option
  * Add `run` option to add to bundle as well as 'require'
  * Add plugin support

4.1.0 / 2014-10-02
==================

  * Allow options to be passed to transforms

4.0.0 / 2014-10-01
==================

  * Update browserify to [5.12.1](https://github.com/substack/node-browserify/blob/master/changelog.markdown#5121)
  * Breaks the `expose` option.  If you need this, don't upgrade yet.

3.0.1 / 2014-07-15
==================

  * Update browserify to [4.2.1](https://github.com/substack/node-browserify/blob/master/changelog.markdown#421) (contains security fix!)

3.0.0 / 2014-05-20
==================

  * Update browserify to [4.1.5](https://github.com/substack/node-browserify/blob/master/changelog.markdown#415) ([N.B. there are some small breaking changes](https://github.com/substack/node-browserify/blob/master/doc/changelog/4_0.markdown)
  * Remove magic path resolution (see #21)

2.8.0 / 2014-05-20
==================

  * Deprecate magic path resolution (see #21)

2.7.0 / 2014-05-20
==================

  * Update browserify to 3.46.1

2.6.1 / 2014-05-06
==================

  * Fix reference error when compiling modules with the `precompile` option (true by default in production)

2.6.0 / 2014-05-02
==================

  * Add `precompile` option for all but directory serving (and enable by default in production)
  * Define some hooks for modifying the entire bundle as a string before and after minifying
  * Update browserify to 3.44.2

2.5.0 / 2014-03-29
==================

  * Update browserify to 3.38.0

2.4.0 / 2014-03-15
==================

  * Update browserify to 3.32.1

2.3.0 / 2014-01-23
==================

  * Update browserify to 3.24.1
  * Fix a bug when the `external` option combined with a "dynamic" `cache`

2.2.0 / 2013-12-16
==================

  * Update browserify to 3.4.0
  * Pass minify options to uglify-js (suggested by [@rnikitin](https://github.com/rnikitin))
  * Pass `resolve` and `basedir` options to browserify

2.1.0 / 2013-12-13
==================

  * Update browserify to 3.2.1

2.0.0 / 2013-12-09
==================

  * Update browserify to 3.0.0 (brings major reductions in file sizes for many browserified modules, deprecates support for some features in older browsers)
  * Suport per-module options to allow custom expose names (thanks to [@KylePDavis](https://github.com/KylePDavis))

1.21.0 / 2013-11-30
===================

  * Update browserify to 2.36.0

1.20.0 / 2013-11-22
===================

  * Update browserify to 2.35.0
  * Update mold-source-map to 0.3.0
  * Refactor (thanks to [duereg](https://github.com/duereg))

1.19.0 / 2013-09-30
===================

  * Update browserify to 2.33.0
  * Support options.extensions (thanks to [@refractalize](https://github.com/refractalize))

1.18.0 / 2013-09-23
===================

  * Update browserify to 2.32.2

1.17.2 / 2013-08-25
===================

  * Update browserify to 2.29.0

1.17.1 / 2013-08-25
===================

  * Update browserify to 2.28.0
  * Update uglify-js to 2.4.0

1.17.0 / 2013-08-05
===================

  * Update browserify to 2.26.0
  * You can serve non-js files via the `grep` option (thanks to [@maxnordlund](https://github.com/maxnordlund))
  * Fix double compile with dynamic caching (thanks to [@ef4](https://github.com/ef4))
  * Fix noParse for bare module names (thanks to [@ef4](https://github.com/ef4))

1.16.0 / 2013-07-25
===================

  * Update browserify to 2.25.1

1.15.0 / 2013-07-25
===================

  * Give source-maps the correct path separators on windows (which leads to a proper folder browser in devtools)
  * Add a `basedir` option to display source map paths relative to

1.14.1 / 2013-07-09
===================

  * Update browserify to 2.24.2

1.14.0 / 2013-07-07
===================

  * Implement dynamic caching (`{cache: 'dynamic'}` is now the default when in development mode)
  * Update browserify to 2.23.1

1.13.1 / 2013-07-03
===================

  * Add support for using without express (except serving a directory)
  * Fix `opts.noParse` (thanks to [@haimschindler](https://github.com/haimschindler))

1.13.0 / 2013-06-26
===================

  * Update browserify to 2.22.0

1.12.0 / 2013-05-31
===================

  * Update browserify to 2.17.2

1.11.0 / 2013-05-30
===================

  * Update browserify to 2.16.0
  * Add npm badge

1.10.0 / 2013-05-21
===================

  * Update browserify to 2.14.1
  * Added `noParse` option

1.9.0 / 2013-05-02
==================

  * Update uglify-js to 2.3.0
  * Added `.npmignore`

1.8.3 / 2013-04-30
==================

  * Fix reservedKeys logic for settings (thanks to [@mhart](https://github.com/mhart))

1.8.2 / 2013-04-30
==================


lib/settings.js
var ms = require('ms');

var reservedKeys = ['normalize', 'env', 'mode'];

exports = module.exports = setter();
exports.env = env;
exports.mode = process.env.NODE_ENV || 'development';


function env(name) {
  if (exports[name]) {
    return exports[name];
  }
  reservedKeys.push(name);
  return exports[name] = setter();
}
function setter(obj) {
  obj = obj || set;
  function set(key) {
    if (arguments.length === 2) {
      obj[key] = arguments[1];
      return this;
    } else if (typeof key === 'object') {
      Object.keys(key)
        .forEach(function (k) {
          obj[k] = key[k];
        });
      return this;
    } else {
      return obj[key];
    }
  }
  return set;
}

exports.external = [];
exports.ignore = [];
exports.ignoreMissing = false;
exports.transform = [];
exports.insertGlobals = false;
exports.detectGlobals = true;
exports.standalone = false;
exports.noParse = [];
exports.extensions = [];
exports.basedir = false;
exports.grep = /\.js$/;

//set some safe defaults for
//unknown environemnts
exports.cache = false;
exports.minify = false;
exports.gzip = true;
exports.debug = false;

var production = exports.env('production');
production.cache = true;
production.precompile = true;
production.minify = true;
production.gzip = true;
production.debug = false;

var development = exports.env('development');
development.cache = 'dynamic';
development.precompile = false;
development.minify = false;
development.gzip = false;
development.debug = true;

exports.normalize = normalize;


function normalize(options) {
  var defaults = exports[exports.mode] || {};
  options = options || {};

  Object.keys(defaults)
    .forEach(function (key) {
      if (options[key] === null || options[key] === undefined) {
        options[key] = defaults[key];
      }
    });
  Object.keys(exports)
    .forEach(function (key) {
      if (reservedKeys.indexOf(key) === -1 && (options[key] === null || options[key] === undefined)) {
        options[key] = exports[key];
      }
    });


  if (options.cache === 'dynamic') {
    // leave unchanged
  } else if (typeof options.cache === 'string' && ms(options.cache)) {
    options.cache = 'public, max-age=' + Math.floor(ms(options.cache)/1000);
  } else if (options.cache === true) {
    options.cache = 'public, max-age=60';
  } else if (typeof options.cache === 'number') {
    options.cache = 'public, max-age=' + Math.floor(options.cache/1000);
  } else if (typeof options.cache === 'object') {
    options.cache = (options.cache.private ? 'private' : 'public') + ', max-age='
                  + Math.floor(ms(options.cache.maxAge.toString())/1000);
  }

  options.precompile = !!options.precompile;
  options.external = arrayify(options.external);
  options.ignore = arrayify(options.ignore);
  options.transform = arrayify(options.transform);
  options.noParse = arrayify(options.noParse);
  options.extensions = arrayify(options.extensions);

  return options;
}

function arrayify(val) {
  if (val && !Array.isArray(val) && typeof val !== 'boolean') {
    return [val];
  } else {
    return val;
  }
}
